File: concepts/10_time_locale_and_runtime/04_runtime_capabilities_versions_and_feature_detection/lua/test_02_interface_and_implementation_layers.lua
-- Common question: which observations belong to the language interface versus one implementation?
-- Inputs: Lua/C release strings, numeric configuration, package separators, bytecode, and table capacity hints.
-- Observations: locked public macros, runtime feature APIs, platform configuration, and unstable representations.
-- polyglot-family: time_locale_and_runtime
-- polyglot-concept: runtime_capabilities_versions_and_feature_detection
-- polyglot-related: languages/lua/tooling_and_runtime/16_advanced_c_api_and_compatibility/
-- polyglot-related+: test_128_api_version_and_abi_boundaries.lua

local t = require("support.assertions")
local c_api = require("support.c_api")
local native = require("polyglot_native")

t.equal(_VERSION, "Lua 5.5")
t.equal(native.release, "Lua 5.5.0")
t.matches(c_api.run("version-gc"), "C API case passed: version%-gc")
t.equal(math.type(1), "integer")
t.equal(string.packsize("j"), 8)
t.equal(package.config:sub(1, 1), "/")
t.equal(type(table.create), "function")

local bytecode = string.dump(function() return 42 end)
t.equal(bytecode:sub(1, 4), "\27Lua")

-- ABI、bytecode、对象布局、GC 时机和 table 内存布局不是跨版本语言保证。
t.done()

File: concepts/10_time_locale_and_runtime/04_runtime_capabilities_versions_and_feature_detection/ruby/test_02_interface_and_implementation_layers.rb
# 共同问题：哪些观察属于 Ruby 接口，哪些只属于锁定 CRuby 实现。
# 输入：语言常量、RubyVM、RbConfig、ObjectSpace 和 C Extension；观察：portable API、实现 namespace 及 ABI 配置。
# polyglot-family: time_locale_and_runtime
# polyglot-concept: runtime_capabilities_versions_and_feature_detection
# polyglot-related: languages/ruby/tooling_and_runtime/16_advanced_runtime_and_compat/
# polyglot-related+: test_127_public_c_abi_configuration.rb

require "assertions"
require "objspace"
require "polyglot_native"
require "rbconfig"

A = PolyglotAssertions

A.equal("ruby", RUBY_ENGINE)
A.equal("4.0.6", RUBY_VERSION)
A.truth(RubyVM.is_a?(Module))
A.equal("aarch64-linux", RbConfig::CONFIG.fetch("arch"))
A.equal("so", RbConfig::CONFIG.fetch("DLEXT"))
A.equal("CRuby 4.0.6", PolyglotNative::RELEASE)

dump = ObjectSpace.dump(Object.new)
A.truth(dump.start_with?("{"))
# RubyVM、ObjectSpace dump 和具体架构只作 CRuby 观察，不上升为跨实现语言保证。
A.truth(defined?(RubyVM::InstructionSequence))

A.done
